### ISSUE TITLE
Improve semver compliance in auto-versioned builds

### ISSUE THREAD

**Issue by ziz** [tribesthatmaybe/modpack#22]
Followup to #19: The commit hash will not strictly increase and therefore doesn't satisfy semver precedence.  

Should the hash be build-metadata rather than pre-release (so `ttmb-client-0.0.1+1c0ce837` instead of `ttmb-client-0.0.1-1c0ce837`)?   Or, even better, we could make the pre-release the date (`ttmb-client-0.0.1-alpha.20200630174612+1c0ce836`) to let precedence work properly on autogenerated non-tagged builds.

(And once precedence is working correctly, we should bump the patch so all future autobuilds will be, to semver, more recent than anything 0.0.1 including non-semver-compliant artifacts)

**Comment by otakup0pe** [comment]
From [semver.org](https://semver.org/) section 9


> A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version. Identifiers MUST comprise only ASCII alphanumerics and hyphens [0-9A-Za-z-]. Identifiers MUST NOT be empty. Numeric identifiers MUST NOT include leading zeroes. Pre-release versions have a lower precedence than the associated normal version. A pre-release version indicates that the version is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version. Examples: 1.0.0-alpha, 1.0.0-alpha.1, 1.0.0-0.3.7, 1.0.0-x.7.z.92, 1.0.0-x-y-z.–.

Nothing in here about pre-release / pre-build metadata needing to increase?

I agree that next steps should include a more clear alpha/beta distinction. That should come slightly before we start yeeting things into curseforge as non-pre-release artifacts built from the `mainline` branch.

**Comment by ziz** [comment]
(Section 11)[https://semver.org/#spec-item-11] describes the precedence and
ordering of prerelease versions:

```Precedence for two pre-release versions with the same major, minor, and
patch version MUST be determined by comparing each dot separated identifier
from left to right until a difference is found as follows:

Identifiers consisting of only digits are compared numerically.

Identifiers with letters or hyphens are compared lexically in ASCII sort
order.

Numeric identifiers always have lower precedence than non-numeric
identifiers.

A larger set of pre-release fields has a higher precedence than a smaller
set, if all of the preceding identifiers are equal.

Example: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta <
1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0.
```

So if it is not strictly increasing the precedence of later prerelease
versions will not be strictly higher than earlier prerelease versions.

(I was not intending the inclusion of `alpha` in this proposal to be
majorly meaning loaded, just "it obviously namespaces the autogen
prereleases to lower priority than explicitly-beta prereleases when we get
there")

**Comment by ziz** [comment]
[Section 11](https://semver.org/#spec-item-11]) describes the precedence and
ordering of prerelease versions:

```
Precedence for two pre-release versions with the same major, minor, and
patch version MUST be determined by comparing each dot separated identifier
from left to right until a difference is found as follows:

Identifiers consisting of only digits are compared numerically.

Identifiers with letters or hyphens are compared lexically in ASCII sort
order.

Numeric identifiers always have lower precedence than non-numeric
identifiers.

A larger set of pre-release fields has a higher precedence than a smaller
set, if all of the preceding identifiers are equal.

Example: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta <
1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0.
```

So if it is not strictly increasing the precedence of later prerelease
versions will not be strictly higher than earlier prerelease versions.

(I was not intending the inclusion of `alpha` in this proposal to be
majorly meaning loaded, just "it obviously namespaces the autogen
prereleases to lower priority than explicitly-beta prereleases when we get
there")

**Comment by hunnybear** [comment]
> Should the hash be build-metadata rather than pre-release (so `ttmb-client-0.0.1+1c0ce837` instead of `ttmb-client-0.0.1-1c0ce837`)? Or, even better, we could make the pre-release the date (`ttmb-client-0.0.1-alpha.20200630174612+1c0ce836`) to let precedence work properly on autogenerated non-tagged builds.

I definitely prefer the latter as well. Anything in the build section (after `+`) is ignored for precedence, this is exactly what I ended up setting up for our team at work

**Comment by otakup0pe** [comment]
aite so i think this can be handled by using the `build` instead of `pre-build` flag for `avakas`. Slightly longer term (bc I'll need to mod avakas) we can add a `alpha` or `dev` prefix.

**Comment by otakup0pe** [comment]
revived `avakas` so gonna do this. will have it include the date as a prefix in prebuild/build meta

also found this gem in the readme

> It also supports the --pre-build argument, which does the same thing on top of the prerelease field, because all kinds of package management systems do not actually support the build semantic version component.

i wonder what tool i was mad at all those years ago 🤷‍♂️ 

https://github.com/otakup0pe/avakas/issues/26

**Comment by otakup0pe** [comment]
![semver artifacts](https://user-images.githubusercontent.com/254586/89704958-df288b80-d90d-11ea-98a5-0e3c0c2e5910.png)

[in this build](https://github.com/tribesthatmaybe/modpack/actions/runs/200171234)
